docs/reference/protocol/typedesc.rst

```diff
@@ -31,83 +31,71 @@ to represent the input descriptor when a query does not receive any arguments,
 or the state descriptor for an empty/default state.
 
 
+Descriptor and type IDs
+=======================
+
+The descriptor and type IDs in structures below are intended to be semi-stable
+unique identifiers of a type.  Fundamental types have globally stable known
+IDs, and type IDs for schema-defined types (i.e. with
+``schema_defined = true``) persist.  Ephemeral type ids are derived from
+type structure and are not guaranteed to be stable, but are still useful
+as cache keys.
+
+
 Set Descriptor
 ==============
 
 .. code-block:: c
 
     struct SetDescriptor {
         // Indicates that this is a Set value descriptor.
-        uint8   type = 0;
+        uint8   tag = 0;
 
         // Descriptor ID.
         uuid    id;
 
         // Set element type descriptor index.
-        uint16  type_pos;
+        uint16  type;
     };
 
 Set values are encoded on the wire as
 :ref:`single-dimensional arrays <ref_protocol_fmt_array>`.
 
 
-Object Shape Descriptor
-=======================
+Scalar Type Descriptor
+======================
 
 .. code-block:: c
 
-    struct ObjectShapeDescriptor {
-        // Indicates that this is an
-        // Object Shape descriptor.
-        uint8           type = 1;
-
-        // Descriptor ID.
-        uuid            id;
-
-        // Number of elements in shape.
-        uint16          element_count;
-
-        ShapeElement    elements[element_count];
-    };
-
-    struct ShapeElement {
-        // Field flags:
-        //   1 << 0: the field is implicit
-        //   1 << 1: the field is a link property
-        //   1 << 2: the field is a link
-        uint32          flags;
-
-        uint8<Cardinality> cardinality;
-
-        // Field name.
-        string          name;
-
-        // Field type descriptor index.
-        uint16          type_pos;
-    };
-
-.. eql:struct:: edb.protocol.enums.Cardinality
+    struct ScalarTypeDescriptor {
+        // Indicates that this is a
+        // Scalar Type descriptor.
+        uint8   tag = 3;
 
-Objects are encoded on the wire as :ref:`tuples <ref_protocol_fmt_tuple>`.
+        // Schema type ID.
+        uuid    id;
 
+        // Schema type name.
+        string  name;
 
-Base Scalar Type Descriptor
-===========================
+        // Whether the type is defined in the schema
+        // or is ephemeral.
+        bool    schema_defined;
 
-.. code-block:: c
+        // Index of fundamental base scalar type
+        // if this is not a fundamental type itself.
+        uint16  fundamental_type;
 
-    struct BaseScalarTypeDescriptor {
-        // Indicates that this is an
-        // Base Scalar Type descriptor.
-        uint8           type = 2;
+        // Number of ancestor scalar types.
+        uint16  ancestors_count;
 
-        // Descriptor ID.
-        uuid            id;
+        // Indexes of ancestor scalar type descriptors
+        // in ancestor resolution order (C3).
+        uint16  ancestors[ancestors_count];
```

Comment:
Do we want there to be a way to distinguish direct parent types?
(Not sure if we actually care but ti seems a little odd to expose ancestors but not direct parents?)

Reply:
I'm not sure it's useful.  If the goal is to apply certain policy based on inheritance, then mro is a lot more useful (you just iterate on the list until you find a match).

Reply:
Oh, yeah, that is true. You can get the exact bases list, since in C3 the ancestor list order must be consistent with the direct bases list order (... not enforcing this was the C3 bug we had).
And the algorithm to get bases from ancestors is simpler than the other way around, so

Reply:
Oh, wait, it's not true that you can do it exactly; you could miss a direct parent that is also an ancestor of an earlier parent. 
But also probably nobody cares